src/project/internal/audacityproject.cpp

```diff
@@ -266,20 +272,28 @@ Ret Audacity4Project::doSave(const muse::io::path_t& savePath, bool generateBack
 
     UNUSED(generateBackup);
 
-    if (createThumbnail) {
-        m_captureThumbnailRequested.notify();
-    }
-
     if ((fileSystem()->exists(savePath) && !fileSystem()->isWritable(savePath))) {
         LOGE() << "failed save, not writable path: " << savePath;
         return make_ret(io::Err::FSWriteError);
     }
 
     auto ret = m_au3Project->save(savePath);
-    if (ret) {
-        return make_ret(Ret::Code::Ok);
+    if (!ret) {
+        return make_ret(Ret::Code::UnknownError);
     }
-    return muse::make_ret(muse::Ret::Code::UnknownError);
+
+    if (createThumbnail) {
+        std::string thumbnailPath = savePath.toStdString();
+        thumbnailPath.replace(thumbnailPath.size() - 4, 4, "png");
+
+        Ret ret = thumbnailCreator()->createThumbnail(thumbnailPath);
```

Comment:
There are two things that bother me here: 
1. What if a path without an extension is passed here, or with some other extension, something like `file.au.backup`... I think it's better to use `muse::io::FileInfo` for path parsing 
2. the fact that the thumbnail is saved in PNG format is something only the creator “knows”, i.e. we either reveal the knowledge here, or we tell the creator to save it in PNG, and not in any other way.

Reply:
Ok. I have created a helper method and it is way better now.